### PR TITLE
[Merged by Bors] - feat: port Algebra.BigOperators.Finsupp

### PR DESCRIPTION
**Pull request by LukasMias**
Some simpNF issues & some leaky classicals caused trouble. But done now!

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/BigOperators/Finsupp.lean [hunk 1]
```diff
@@ -59,107 +60,118 @@ variable [Zero M] [Zero M'] [CommMonoid N]
...
 theorem prod_ite_eq [DecidableEq α] (f : α →₀ M) (a : α) (b : α → M → N) :
-    (f.Prod fun x v => ite (a = x) (b x v) 1) = ite (a ∈ f.support) (b a (f a)) 1 := by
+    (f.prod fun x v => ite (a = x) (b x v) 1) = ite (a ∈ f.support) (b a (f a)) 1 := by
   dsimp [Finsupp.prod]
   rw [f.support.prod_ite_eq]
 #align finsupp.prod_ite_eq Finsupp.prod_ite_eq
 #align finsupp.sum_ite_eq Finsupp.sum_ite_eq
 
-@[simp]
+-- Porting note: simpnf linter; Left-hand side simplifies from
```
Comment (jcommelin) [inline Mathlib/Algebra/BigOperators/Finsupp.lean:117]: Please add a new theorem `sum_ite_self_eq_aux` that prove that the simplified LHS is equal to the RHS and tag it with `simp`. That way we maintain the same simp-set as mathlib 3.

File: Mathlib/Algebra/BigOperators/Finsupp.lean [hunk 2]
```diff
@@ -59,107 +60,118 @@ variable [Zero M] [Zero M'] [CommMonoid N]
...
 theorem prod_ite_eq' [DecidableEq α] (f : α →₀ M) (a : α) (b : α → M → N) :
-    (f.Prod fun x v => ite (x = a) (b x v) 1) = ite (a ∈ f.support) (b a (f a)) 1 := by
+    (f.prod fun x v => ite (x = a) (b x v) 1) = ite (a ∈ f.support) (b a (f a)) 1 := by
   dsimp [Finsupp.prod]
   rw [f.support.prod_ite_eq']
 #align finsupp.prod_ite_eq' Finsupp.prod_ite_eq'
 #align finsupp.sum_ite_eq' Finsupp.sum_ite_eq'
 
-@[simp]
+-- Porting note: simpnf linter; Left-hand side simplifies from
```
Comment (jcommelin) [inline Mathlib/Algebra/BigOperators/Finsupp.lean:138]: Idem dito.

File: Mathlib/Algebra/BigOperators/Finsupp.lean [hunk 3]
```diff
@@ -564,47 +600,47 @@ theorem Finset.sum_apply' : (∑ k in s, f k) i = ∑ k in s, f k i :=
...
-theorem Finsupp.sum_apply' : g.Sum k x = g.Sum fun i b => k i b x :=
+theorem Finsupp.sum_apply' : g.sum k x = g.sum fun i b => k i b x :=
   Finset.sum_apply _ _ _
 #align finsupp.sum_apply' Finsupp.sum_apply'
 
 section
 
-include h0 h1
+-- Porting note: removed the following, expected command
+-- include h0 h1
```
Comment (jcommelin) [inline Mathlib/Algebra/BigOperators/Finsupp.lean:610]:
```suggestion
```
This is expected.

File: Mathlib/Algebra/BigOperators/Finsupp.lean [hunk 4]
```diff
@@ -564,47 +600,47 @@ theorem Finset.sum_apply' : (∑ k in s, f k) i = ∑ k in s, f k i :=
...
 
-theorem Finsupp.sum_sum_index' : (∑ x in s, f x).Sum t = ∑ x in s, (f x).Sum t :=
+theorem Finsupp.sum_sum_index' : (∑ x in s, f x).sum t = ∑ x in s, (f x).sum t :=
   Finset.induction_on s rfl fun a s has ih => by
     simp_rw [Finset.sum_insert has, Finsupp.sum_add_index' h0 h1, ih]
 #align finsupp.sum_sum_index' Finsupp.sum_sum_index'
 
-end
+-- Porting note: unknown namespace end
+-- end
```
Comment (jcommelin) [inline Mathlib/Algebra/BigOperators/Finsupp.lean:620]: This is a bit weird. I would expect this to end the section started on L607.

Comment (LukasMias) [reply Mathlib/Algebra/BigOperators/Finsupp.lean]: There used to be an error here but I don't see it anymore. Maybe hallucinations. Thanks!

File: Mathlib/Algebra/BigOperators/Finsupp.lean [hunk 5]
```diff
@@ -0,0 +1,649 @@
...
+theorem coe_sum [Zero M] [AddCommMonoid N] (f : α →₀ M) (g : α → M → β →₀ N) :
+    ⇑(f.sum g) = f.sum fun a₁ b => ⇑(g a₁ b) :=
+  coe_finset_sum _ _
+#align finsupp.coe_sum Finsupp.coe_sum
+
+theorem support_sum [DecidableEq β] [Zero M] [AddCommMonoid N] {f : α →₀ M} {g : α → M → β →₀ N} :
+    (f.sum g).support ⊆ f.support.bunionᵢ fun a => (g a (f a)).support := by
+  have : ∀ c, (f.sum fun a b => g a b c) ≠ 0 → ∃ a, f a ≠ 0 ∧ ¬(g a (f a)) c = 0 := fun a₁ h =>
+    let ⟨a, ha, Ne⟩ := Finset.exists_ne_zero_of_sum_ne_zero h
+    ⟨a, mem_support_iff.mp ha, Ne⟩
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/BigOperators/Finsupp.lean:335]:
```suggestion
    let ⟨a, ha, ne⟩ := Finset.exists_ne_zero_of_sum_ne_zero h
    ⟨a, mem_support_iff.mp ha, ne⟩
```

File: Mathlib/Algebra/BigOperators/Finsupp.lean [hunk 6]
```diff
@@ -0,0 +1,649 @@
...
+  Multiset.induction_on s (single_zero _) fun a s ih => by
+    rw [Multiset.sum_cons, single_add, ih, Multiset.map_cons, Multiset.sum_cons]
+#align finsupp.single_multiset_sum Finsupp.single_multiset_sum
+
+theorem single_finset_sum [AddCommMonoid M] (s : Finset ι) (f : ι → M) (a : α) :
+    single a (∑ b in s, f b) = ∑ b in s, single a (f b) := by
+  trans
+  apply single_multiset_sum
+  rw [Multiset.map_map]
+  rfl
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/BigOperators/Finsupp.lean:289]:
```suggestion
  · apply single_multiset_sum
  · rw [Multiset.map_map]
    rfl
```

File: Mathlib/Algebra/BigOperators/Finsupp.lean [hunk 7]
```diff
@@ -0,0 +1,649 @@
...
+    f.sum.sum h = (f.map fun g : α →₀ M => g.sum h).sum :=
+  Multiset.induction_on f rfl fun a s ih => by
+    rw [Multiset.sum_cons, Multiset.map_cons, Multiset.sum_cons, sum_add_index' h₀ h₁, ih]
+#align finsupp.multiset_sum_sum_index Finsupp.multiset_sum_sum_index
+
+theorem support_sum_eq_bunionᵢ {α : Type _} {ι : Type _} {M : Type _} [DecidableEq α]
+    [AddCommMonoid M] {g : ι → α →₀ M} (s : Finset ι)
+    (h : ∀ i₁ i₂, i₁ ≠ i₂ → Disjoint (g i₁).support (g i₂).support) :
+    (∑ i in s, g i).support = s.bunionᵢ fun i => (g i).support := by
+  classical have : DecidableEq ι := inferInstance
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/BigOperators/Finsupp.lean:549]:
```suggestion
  classical
```

File: Mathlib/Algebra/BigOperators/Finsupp.lean [hunk 8]
```diff
@@ -0,0 +1,649 @@
...
+#align finset.sum_apply' Finset.sum_apply'
+
+theorem Finsupp.sum_apply' : g.sum k x = g.sum fun i b => k i b x :=
+  Finset.sum_apply _ _ _
+#align finsupp.sum_apply' Finsupp.sum_apply'
+
+section
+
+-- include h0 h1
+
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/BigOperators/Finsupp.lean:615]:
```suggestion
```

File: Mathlib/Algebra/BigOperators/Finsupp.lean [hunk 9]
```diff
@@ -0,0 +1,649 @@
...
+#align finsupp.mul_sum Finsupp.mul_sum
+
+end
+
+namespace Nat
+
+-- Porting note: Needed to replace pow with Pow.pow
+/-- If `0 : ℕ` is not in the support of `f : ℕ →₀ ℕ` then `0 < ∏ x in f.support, x ^ (f x)`. -/
+theorem prod_pow_pos_of_zero_not_mem_support {f : ℕ →₀ ℕ} (hf : 0 ∉ f.support) :
+    0 < f.prod Pow.pow :=
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/BigOperators/Finsupp.lean:644]: I think we prefer
```suggestion
    0 < f.prod (· ^ ·) :=
```

**Review by Ruben-VandeVelde — APPROVED** [review]

**Pull request closed without merging**